lib/src/dag_walk.rs

```diff
@@ -14,54 +14,32 @@
 
 use std::collections::HashSet;
 use std::hash::Hash;
-use std::iter::Iterator;
+use std::iter;
 
-pub struct BfsIter<'id_fn, 'neighbors_fn, T, ID, NI> {
-    id_fn: Box<dyn Fn(&T) -> ID + 'id_fn>,
-    neighbors_fn: Box<dyn FnMut(&T) -> NI + 'neighbors_fn>,
-    work: Vec<T>,
-    visited: HashSet<ID>,
-}
-
-impl<T, ID, NI> Iterator for BfsIter<'_, '_, T, ID, NI>
-where
-    ID: Hash + Eq,
-    NI: IntoIterator<Item = T>,
-{
-    type Item = T;
-
-    fn next(&mut self) -> Option<Self::Item> {
-        loop {
-            let c = self.work.pop()?;
-            let id = (self.id_fn)(&c);
-            if self.visited.contains(&id) {
-                continue;
-            }
-            for p in (self.neighbors_fn)(&c) {
-                self.work.push(p);
-            }
-            self.visited.insert(id);
-            return Some(c);
-        }
-    }
-}
-
-pub fn bfs<'id_fn, 'neighbors_fn, T, ID, II, NI>(
+pub fn bfs<T, ID, II, NI>(
     start: II,
-    id_fn: Box<dyn Fn(&T) -> ID + 'id_fn>,
-    neighbors_fn: Box<dyn FnMut(&T) -> NI + 'neighbors_fn>,
-) -> BfsIter<'id_fn, 'neighbors_fn, T, ID, NI>
+    id_fn: impl Fn(&T) -> ID,
+    mut neighbors_fn: impl FnMut(&T) -> NI,
+) -> impl Iterator<Item = T>
 where
     ID: Hash + Eq,
     II: IntoIterator<Item = T>,
     NI: IntoIterator<Item = T>,
 {
-    BfsIter {
-        id_fn,
-        neighbors_fn,
-        work: start.into_iter().collect(),
-        visited: Default::default(),
-    }
+    let mut work: Vec<T> = start.into_iter().collect();
+    let mut visited: HashSet<ID> = HashSet::new();
+    iter::from_fn(move || loop {
+        let c = work.pop()?;
+        let id = id_fn(&c);
+        if visited.contains(&id) {
+            continue;
+        }
+        for p in neighbors_fn(&c) {
+            work.push(p);
+        }
+        visited.insert(id);
```

Comment:
Usually, when I code BFS, I use `visited` to avoid inserting the same node several times in `work` (which you discard later), I think that this is more efficient if there are many duplicates as it hashes less `ID` instances. This would give something like (untested):

```rust
let mut visited = work.iter().map(id_fn).collect::<HashSet<_>>();
…
iter::from_fn(move || {
  let c = work.pop()?;
  neighbors_fn(&c)
    .into_iter()
    .filter(|p| visited.insert(id_fn(p)))
    .for_each(|p| work.push(p));
  Some(c)
})
```

Reply:
renamed it to `dfs()`